### DISCUSSION TITLE
Problem relevnt to PatchMerging in 3D SWIN-UNETR

### DISCUSSION THREAD

**Discussion by GYDDHPY** [Project-MONAI/MONAI#4753] [Q&A] [answered]
In the 3D forwrd function of PatchMerging, it seems x3 equals to x6, which is a little bit beird.
Is this a special design for 3D image or just a code error?

https://github.com/Project-MONAI/MONAI/blob/398466c28adda2af25ed931786f22af1ea74483f/monai/networks/nets/swin_unetr.py#L660-L701

**Comment by tangy5** [comment, marked as answer]
Hi Thanks for pointing out this issue. This seems to be a design option, and also a code typo here. 
This issue originate whether to merge and downsample along Z axis, initially, it was
```
        x0 = x[:, :, 0::2, 0::2, :]  # B D H/2 W/2 C
        x1 = x[:, :, 1::2, 0::2, :]  # B D H/2 W/2 C
        x2 = x[:, :, 0::2, 1::2, :]  # B D H/2 W/2 C
        x3 = x[:, :, 1::2, 1::2, :]  # B D H/2 W/2 C
        x = torch.cat([x0, x1, x2, x3], -1)  # B D H/2 W/2 4*C
```
And yes, after discussion on the 3D, the patch merging should be:
```diff
            x0 = x[:, 0::2, 0::2, 0::2, :]
            x1 = x[:, 1::2, 0::2, 0::2, :]
            x2 = x[:, 0::2, 1::2, 0::2, :]
            x3 = x[:, 0::2, 0::2, 1::2, :]
            x4 = x[:, 1::2, 0::2, 1::2, :]
-           x5 = x[:, 0::2, 1::2, 0::2, :]
-           x6 = x[:, 0::2, 0::2, 1::2, :]
+           x5 = x[:, 1::2, 1::2, 0::2, :]
+           x6 = x[:, 0::2, 1::2, 1::2, :]
            x7 = x[:, 1::2, 1::2, 1::2, :]
            x = torch.cat([x0, x1, x2, x3, x4, x5, x6, x7], -1)
```

We added patchMergingV2 as follow:
https://github.com/Project-MONAI/MONAI/pull/4758/commits/a4ef630fcd24e0628e396347bbcb3c7cd1818652

Thanks again

**Reply by GYDDHPY** [reply]
Got it!
Thank you for your attention.

**Comment by function2-llx** [comment]
Maybe using a for loop is less error-prone for situation like this, I think XD

Something like:

```python
x = torch.cat([
	x[:, i::2, j::2, k::2, :] 
	for i, j, k in itertools.product(range(2), range(2), range(2))
], -1)
```

**Reply by wyli** [reply]
Yes, it also improves the readability, are you interested in submitting a PR? @function2-llx

**Reply by function2-llx** [reply]
With pleasure!
It's here: https://github.com/Project-MONAI/MONAI/pull/4793/
